map/local_ads_manager.cpp

```diff
@@ -0,0 +1,350 @@
+
+  return result;
+}
+
+std::chrono::steady_clock::time_point DeserializeTimestamp(ModelReaderPtr const & reader)
+{
+  ASSERT_EQUAL(reader.Size(), sizeof(long), ());
+
+  std::vector<uint8_t> bytes(reader.Size());
+  reader.Read(0, bytes.data(), bytes.size());
+
+  MemReaderWithExceptions memReader(bytes.data(), bytes.size());
+  ReaderSource<decltype(memReader)> src(memReader);
+  long hours = ReadPrimitiveFromSource<long>(src);
+
+  return std::chrono::steady_clock::time_point(std::chrono::hours(hours));
+}
+
+std::string GetPath(std::string const & fileName)
+{
+  return my::JoinFoldersToPath(GetPlatform().WritableDir(), fileName);
+}
+}  // namespace
+
+LocalAdsManager::LocalAdsManager(GetMwmsByRectFn const & getMwmsByRectFn,
+                       GetMwmIdByName const & getMwmIdByName)
+  : m_getMwmsByRectFn(getMwmsByRectFn)
+  , m_getMwmIdByNameFn(getMwmIdByName)
+  , m_thread(&LocalAdsManager::ThreadRoutine, this)
+{
+  CHECK(m_getMwmsByRectFn != nullptr, ());
+  CHECK(m_getMwmIdByNameFn != nullptr, ());
+}
+
+LocalAdsManager::~LocalAdsManager()
+{
+#ifdef DEBUG
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    ASSERT(!m_isRunning, ());
+  }
+#endif
+}
+
+void LocalAdsManager::Teardown()
+{
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (!m_isRunning)
+      return;
+    m_isRunning = false;
+  }
+  m_condition.notify_one();
+  m_thread.join();
+}
+
+void LocalAdsManager::SetDrapeEngine(ref_ptr<df::DrapeEngine> engine)
+{
+  m_drapeEngine = engine;
+  {
+    std::lock_guard<std::mutex> lock(m_symbolsCacheMutex);
+    if (!m_symbolsCache.empty())
+      m_drapeEngine->AddCustomSymbols(std::move(m_symbolsCache));
+  }
+}
+
+void LocalAdsManager::UpdateViewport(ScreenBase const & screen)
+{
+  auto connectionStatus = GetPlatform().ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_NONE ||
+      df::GetZoomLevel(screen.GetScale()) <= scales::GetUpperWorldScale())
+  {
+    return;
+  }
+
+  std::vector<std::string> requestedCampaigns;
+  auto mwms = m_getMwmsByRectFn(screen.ClipRect());
+  if (mwms.empty())
+    return;
+
+  // Request local ads campaigns.
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    for (auto const & mwm : mwms)
+    {
+      auto info = mwm.GetInfo();
+      if (!info)
+        continue;
+      std::string const & mwmName = info->GetCountryName();
+      auto campaignIt = m_campaigns.find(mwmName);
+      if (campaignIt == m_campaigns.end())
+      {
+        requestedCampaigns.push_back(mwmName);
+        continue;
+      }
+
+      // If a campaign has not been requested from server this session.
+      if (!campaignIt->second)
+      {
+        auto const it = m_expiration.find(mwmName);
+        bool needUpdateByTimeout = (connectionStatus == Platform::EConnectionType::CONNECTION_WIFI);
+        if (!needUpdateByTimeout && it != m_expiration.end())
+        {
+          auto const currentTime = std::chrono::steady_clock::now();
```

Comment:
Ah, sorry for misleading you. We can't use steady_clock as it's not the same as the wall clock. Unfortunately we need to use system clock, and check here that currentTime is greater than it->second.